rfc/src/rfcs/0009-function-contracts.md

```diff
@@ -0,0 +1,845 @@
+lingering threats to soundness in the [future](#future-possibilities) section)
+and guarding against abstractions diverging from their checks.
+
+### Write Sets and Havocking
+
+Functions can have side effects on data reachable through mutable references or
+pointers. To overapproximate all such modifications a function could apply to
+pointed-to data, the verifier "havocs" those regions, essentially replacing
+their content with non-deterministic values.
+
+Let us consider a simple example of a `pop` method.
+
+```rs
+impl<T> Vec<T> {
+  fn pop(&mut self) -> Option<T> {
+    ...
+  }
+}
+```
+
+This function can, in theory, modify any memory behind `&mut self`, so this is
+what Kani will assume it does by default. It infers the "write set", that is the
+set of memory locations a function may modify, from the type system. As a result,
+any data pointed to by a mutable reference or pointer is considered part of the
+write set. In addition, a static analysis of the source code discovers any
+`static mut` variables the function or it's dependencies reference and adds all
+pointed-to data to the write set also. During havocking the verifier replaces
+all locations in the write set with non-deterministic values.
+
+While the inferred write set is sound and enough for successful contract
+checking[^inferred-footprint] in many cases this inference is too coarse
+grained. In the case of `pop` every value in this vector will be made
+non-deterministic.
+
+To address this the proposal also adds a `modifies` and `frees` clause which
+limits the scope of havocking. Both clauses represent an assertion that the
+function will modify only the specified memory regions. Similar to
+requires/ensures the verifier enforces the assertion in the checking stage to
+ensure soundness. When the contract is used as an abstraction, the `modifies`
+clause is used as the write set to havoc.
+
+In our `pop` example the only modified memory location is the last element and
+only if the vector was not already empty, which would be specified thusly.
+
+```rs
+impl<T> Vec<T> {
+  #[modifies(if !self.is_empty() => (*self).buf.ptr.pointer.pointer[self.len])]
+  #[modifies(if self.is_empty())]
+  fn pop(&mut self) -> Option<T> {
+    ...
+  }
+}
+```
+
+The `#[modifies(when = CONDITION, targets = { MODIFIES_RANGE, ... })]` consists
+of an `CONDITION` and zero or more, comma separated `MODIFIES_RANGE`s which are
+essentially an lvalue.
+
+If no `when` is provided the condition defaults to `true`, meaning the modifies
+ranges apply to all invocations of the function. If `targets` is omitted it
+defaults to `{}`, e.g. an empty set of targets meaning under this condition the
+function modifies no mutable memory.
+
+Lvalues are simple expressions permissible on the left hand side of an
+assignment. They compose of the name of one function argument (or static
+variable) and zero or more projections (dereference `*`, field access `.x`,
```

Comment:
Intersting question, I was actually thinking of `*` applying to references, not to pointers, because that's how rust handles it but I think you're right we will need a way to address values behind pointers too. I'm adding it to open questions.

Reply:
good